### ISSUE TITLE
Cannot fetch token from destination service

### ISSUE THREAD

**Issue by FlorianFechterTCon** [SAP/cloud-sdk-js#1963]
Dear Cloud-SDK Team,

last week I tried to redeploy a poc for notifications, my colleague and I created last year.
But I got an error while sending a notification.
After taking a look at your upgrade notes, I realized that there might be some "breaking" changed that happened in the past weeks.
So I upgraded to the 1.53.0, but I still get an error:
```
(node:178) UnhandledPromiseRejectionWarning: ErrorWithCause: Create request failed!
    at new ErrorWithCause (/home/vcap/app/node_modules/@sap-cloud-sdk/core/node_modules/@sap-cloud-sdk/util/dist/error-with-cause.js:34:16)
    at /home/vcap/app/node_modules/@sap-cloud-sdk/core/dist/odata-common/request-builder/create-request-builder-base.js:121:31
    at processTicksAndRejections (internal/process/task_queues.js:95:5)

Caused by:
ErrorWithCause: Could not find a destination with name "notification-endpoint"! Unable to execute request.
    at new ErrorWithCause (/home/vcap/app/node_modules/@sap-cloud-sdk/core/node_modules/@sap-cloud-sdk/util/dist/error-with-cause.js:34:16)
    at /home/vcap/app/node_modules/@sap-cloud-sdk/core/dist/odata-common/request-builder/request-builder-base.js:192:23
    at processTicksAndRejections (internal/process/task_queues.js:95:5)

Caused by:
ErrorWithCause: Could not fetch client credentials token for service of type "destination".
    at new ErrorWithCause (/home/vcap/app/node_modules/@sap-cloud-sdk/core/node_modules/@sap-cloud-sdk/util/dist/error-with-cause.js:34:16)
    at Object.<anonymous> (/home/vcap/app/node_modules/@sap-cloud-sdk/core/dist/connectivity/scp-cf/token-accessor.js:97:27)
    at step (/home/vcap/app/node_modules/@sap-cloud-sdk/core/dist/connectivity/scp-cf/token-accessor.js:44:23)
    at Object.throw (/home/vcap/app/node_modules/@sap-cloud-sdk/core/dist/connectivity/scp-cf/token-accessor.js:25:53)
    at rejected (/home/vcap/app/node_modules/@sap-cloud-sdk/core/dist/connectivity/scp-cf/token-accessor.js:17:65)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)

Caused by:
Error: Request failed with status code 401
    at createError (/home/vcap/app/node_modules/@sap-cloud-sdk/core/node_modules/axios/lib/core/createError.js:16:15)
    at settle (/home/vcap/app/node_modules/@sap-cloud-sdk/core/node_modules/axios/lib/core/settle.js:17:12)
    at IncomingMessage.handleStreamEnd (/home/vcap/app/node_modules/@sap-cloud-sdk/core/node_modules/axios/lib/adapters/http.js:269:11)
    at IncomingMessage.emit (events.js:412:35)
    at endReadableNT (internal/streams/readable.js:1334:12)
    at processTicksAndRejections (internal/process/task_queues.js:82:21)
```
    
But I can not figure out what I did wrong or what I missed when reading the notes.
It would be nice if you folks can help me here.

Some General Information.
- We have a "multi-tenant environment" so one subaccount provides our backend services (that includes the one which wants to send the notification) and another one that provides the launchpad service, notification service, the destination for notifications and is the one we used for authentication.
- Our environment setting for the destination service:
![image](https://user-images.githubusercontent.com/41566727/148752188-f111d16b-59e9-41d3-89ba-686698e5bdce.png)
- Our request call
![image](https://user-images.githubusercontent.com/41566727/148780930-b5f2e755-ab7c-49a7-ba8f-3bf2b383eafe.png)
- Request header send to the authentication service
![image](https://user-images.githubusercontent.com/41566727/148778700-e0de5c72-4b43-46c0-8e7d-438e4a08aa26.png)

**Comment by FrankEssenberger** [comment]
Hi @FlorianFechterTCon,

thanks for reaching out. So what I can see from the logs is that there is an issue for fetching the destination service token (client credentials grant). This token is then used to get the destination from destination. The reason is 401 which indicates that the something is wrong with the service credentials or JWT.

-> I assume you have a XSUAA subscription in the subaccount doing the call?

If this is the case, I guess the quickest way to find the wrong credentials is to duplicate the `VCAP_SERVICES` content locally and put it to the process.env and run this process the `serviceToken()` method directly.  Just provide the service name which is `destination` if I see it right. This methods also takes a JWT but it considers the `iss` and `zid` properties of the JWT so you can also mimic that locally without a full JWT. 

Best
Frank

**Comment by FlorianFechterTCon** [comment]
Hey @FrankEssenberger,

thank you very much for your answer / suggestion. 
I'll try it out today and write you back as soon as I have any news.
About your assumption:
Yes, our subaccount (the one that provides our backend service that tries to send the notification) has a XSUAA subscription

Best regards 

Florian

**Comment by FrankEssenberger** [comment]
When you can receive a token you can directly try one level higher and call the `getDestination()` method. Same rules for the JWT also apply there.

**Comment by FlorianFechterTCon** [comment]
Hey @FrankEssenberger,

it took a little longer to create a small "easy" text case locally.
One thing I realized was that it depended on where I retrieved my token (where authorized).
When I used the "customer" authorization URL/JWT, I could not retrieve a token for the destination service. (401)
On the other hand, if I use the authorization URL/JWT of the providing subaccount, I was able to get the destination service token and use it, but since I defined the destination for the notification-service within the "customer" subaccount, I still could not retrieve it.
I thought the customer (subscribing) subaccount should be allowed to use the destination service, of our providing account. Hmm maybe there is something wrong with my destination-service/btp configuration.
So I'll take a look into those within the next few days.

Best regards

Florian

**Comment by FrankEssenberger** [comment]
Good morning @FlorianFechterTCon,

I guess you have corner the problem well and I also assume the subscribing account is not allowed to call the destination service due to some incorrect config. One thing you can do to check this locally is to create subscriptions to the destination service also in the subscription account and use these credentials in the vcap together with the subscriber jwt. This is of course only for testing.

In the real setup you would create a `dependencies` endpoint in your application. This endpoint is called when you create a subscription to your app in the subscriber account and tells the account which services are consumed by the application. Here a sample code from our E2E test application:
```
export function dependencyRoute(req: Request, res: Response) {
  res.status(200).send(
    relevantServices
      .map(s => {
        const services = EnvironmentAccessor.getServiceList(s);

        return services && services.length
          ? {
              appId:
                services[0].credentials.xsappname ||
                services[0].credentials.uaa.xsappname,
              appName: s
            }
          : null;
      })
      .filter(elem => elem)
  );
}```

You can make this code generic as well that simply all services from the application are passed - however in practice the used service stay rather constant. There are also other endpoint to trigger action on `unsubscribe` but I am no expert there. Here is also a [blog post](https://blogs.sap.com/2020/08/03/multi-tenancy-of-destination-service-with-cloud-foundry-applications/) which covers this topic - I am not sure if this is the best reference but I am sure you find the official documents or other information from there and can adjust the config.

Best
Frank

**Comment by FlorianFechterTCon** [comment]
Hey @FrankEssenberger,

sorry for the late reply, we had some other topics that needed to be tackled first. 😅
And I needed some time to get a better understanding of the topic.
But with the help of some colleagues of mine. We found the problem. 
It was the result of me trying to define the dependencies/service binding on the fly and in between deployments that reverted them. And finally, me forgetting to update the subscribing accounts (that was the root cause) via the subscription management dashboard after all configs were added / deployed. 
Like we said, there was something wrong with the config. 😁

Thank you so much again for your time and support, I'm sorry it was such a small thing. 

Best regards

Florian
